### DISCUSSION TITLE
NullReferenceException in SetProperty or GetProperty when adding row from Telerik grid

### DISCUSSION THREAD

**Discussion by Mikael-Pothier** [MarimerLLC/csla#3201] [Questions]
Hello,

I'm currently updating csla for my blazor project to the latest version(from 5.4) and it went well for most of the update. But when I started testing the site I couldn't add or modify any grid. The grid we are using are from Telerik. When I'm trying to add a row in a grid I have this error in Csla in the SetProperty(it can also be in the GetProperty)
![image](https://user-images.githubusercontent.com/8630076/200923516-af2bcaed-f4a6-4cfb-bba7-848aec8983ae.png)

And I check the value and it is false and the property seems fine too
![image](https://user-images.githubusercontent.com/8630076/200923845-71ec02b6-a7f7-4284-b97c-0afb3b167dd0.png)

Unfortunately, I don't know precisely when the error occurred since it is in the code behind of Telerik and I cannot access it. 

I know it might be a problem from Telerik but I wasn't sure where to start so sorry if I'm not at the right place And thank you for your help in advance.

**Comment by TheCakeMonster** [comment]
Have you got some basic unit tests to exercise your business objects in the same way as within your UI, but disassociated from it? That might be a way to split the problem in half, and escape the potential for it being a problem with the Telerik UI components.

I wonder if it's to do with the way you are creating the objects in the first place. Are you getting the original objects via an injected instance of IDataPortal<T>? Your error reminds me a bit of the type of errors you see when an object was created by newing it up, rather than it having been created by the DataPortal. That might seem obvious, but it's very easy to get wrong.

Do you have any rules associated with the property that is throwing the exception?

There has been some discussion elsewhere about the creation of new child objects in collections. Have a read of issue #3141, which I think resulted in a fix, but I wasn't following closely.

**Reply by Mikael-Pothier** [reply]
Yes, we have some basic test which add a parent and then add a child and everything works properly. The property doesn't have any rules and I tried what was in #3141 and it gives the same error. I added the override method with the part that set the application (see image below) but it doesn't go into this function when I debug my code.
![image](https://user-images.githubusercontent.com/8630076/201118204-5a2a29ec-7049-4a1d-ac20-ee1619ec1abb.png)

We are creating the dataportal via DI in a razor page. Here is the inject of the dataportal and below the call that fetch my parent object which contain my list of child
![image](https://user-images.githubusercontent.com/8630076/201112100-c2fdde54-7317-42df-b73c-e5a82f20d9e1.png)

![image](https://user-images.githubusercontent.com/8630076/201112527-0a1cdbef-17b8-46cb-8d6c-4312f38bc4a0.png)

I think you might be onto something with "creating by newing it up" because I think Telerik create a copy of the object since they have a cancel button that gets back the original object and they probably create it via the constructor.

**Comment by Mikael-Pothier** [comment]
With further test I found out that the ApplicationContext is Null when it goes into the SetProperty. 
![image](https://user-images.githubusercontent.com/8630076/201139130-8d536891-a020-4a83-87c0-424df8b2eafe.png)

Here is the stack trace of the error
at Csla.Core.BusinessBase.VerifyAuthorizationCache()
   at Csla.Core.BusinessBase.CanWriteProperty(IPropertyInfo property)
   at Csla.Core.BusinessBase.CanWriteProperty(IPropertyInfo property, Boolean throwOnFalse)
   at Csla.Core.BusinessBase.SetProperty[P](PropertyInfo`1 propertyInfo, P newValue, NoAccessBehavior noAccess)
   at Primaco.Library.Versements.set_EdiCash(Boolean value) in C:\Users\potml.primdev\Documents\Source\Repos\Pforce\Primaco.Library\Entities\Versements.cs:line 38

If I go in the VerifyAuthorizationCache it does use the applicationContext like this
![image](https://user-images.githubusercontent.com/8630076/201139391-3d41a88d-d91f-45eb-ae44-c086e5207b7d.png)

**Reply by rockfordlhotka** [reply]
If `ApplicationContext` is null when a property is set, that indicates that the business object instance wasn't created correctly. It was probably created with the `new` keywork or a call to `Activator.CreateInstance` - neither of which will set the `ApplicationContext` value as the object is created.

The easiest way to correctly create an instance of a business class is to use `ApplicationContext.CreateInstanceDI` - assuming the calling code where the instance is being created has access to an instance of `ApplicationContext`.

Since this sounds like a child collection, it is probably in the `AddNew` or `AddNewCore` override of the collection class that the child object is being created.

Or, if the datagrid control is directly creating the child via the `new` keyword (which is possible), then things are a bit more complex. If that's the case, I am pretty sure you can override or detect as the child is added to the collection, and in the collection code you implement to handle that case, you can set `ApplicationContext` yourself. Something like this:

```c#
  private void HandleAddingOfNewChild(object newChild)
  {
    if (newChild is IUseApplicationContext iuac)
    {
      iuac.ApplicationContext = this.ApplicationContext;
    }
  }
```

That's basically what `ApplicationContext.CreateInstanceDI` does - creates the instance, then uses the `IUseApplicationContext` interface to set the new object's `ApplicationContext` value.

**Reply by Mikael-Pothier** [reply]
It is in add/edit of the Telerik grid that the problem happened and from what I can see it is not a child at the moment of the error.
![image](https://user-images.githubusercontent.com/8630076/201205867-e4dad5dd-2004-4f8b-8a16-f559c91bdc41.png)

What I think is that Telerik create a copy of the elem in case we want to cancel our modification and they pass by the default constructor. Is there a way to set the applicationContext when it's the parent object?

Also I just want to be sure the code that set the ApplicationContext I need to put it in the back end not in my front end?

**Reply by rockfordlhotka** [reply]
What UI tech is this? WPF? Blazor?

If it is WPF then _maybe_ there's a solution, in that you could expose `ApplicationContext` as a static property off the `App` class, and then in the constructor of the child class you could set its `ApplicationContext` property from the global static.

There's some complexity here, in that you'd need to somehow abstract this code so it doesn't fail on the app server, and I haven't thought through how that would work.

> Also I just want to be sure the code that set the ApplicationContext I need to put it in the back end not in my front end?

I think we're only talking about front-end UI code here, as it is the datagrid control that's creating an instance of the business class and then using that instance before you are able to set its `ApplicationContext`. On the app server everything should be handled by CSLA automatically.

You know, another possibility is that perhaps the datagrid control has an event that you can handle to initialize the new object once it has been created? Do they raise a "row created" or "object created" event maybe?

**Reply by saulyt** [reply]
OP mentioned that it was a Blazor Telerik grid. It looks like there is an [OnModelInit](https://docs.telerik.com/blazor-ui/components/grid/events#onmodelinit) event that can be used to create the object. That probably can be used to create the object via the DataPortal.

**Reply by Mikael-Pothier** [reply]
Hi, sorry for the late reply. The `OnModelInit` worked, but I got another error in Telerik and I waited for their solution before answering here.

The problem was when we add/modify a row the grid creates a copy of the row and it passes in `Activator.CreateInstance` to create it which cause the `ApplicationContext` to be null. Telerik will add a feature so that you can populate the model manually. If anyone are interested in this feature, here is the Telerik ticket https://feedback.telerik.com/blazor/1587593-add-the-ability-to-manually-populate-the-model-properties-when-creating-adding-items

Thank you for your help.

**Reply by mikeclimbs23** [reply]
I'm hitting this issue with Telerik also. Did you ever find a workaround? I was able to handle the add case with OnModelInit, but haven't found a solution for the edit case.

**Reply by fundez** [reply]
I reached out to Telerik about this, They're planning to release a fix to accommodate parameters when creating copies of objects summer of 2024.

Are they any workarounds I can use for parameterless creation while I wait for Telerik to create a fix? We need enhanced grid editing so ditching Telerik isn't a great option for us.

**Reply by MarkOverstreet** [reply]
Just an option for you, DevExpress has worked great for us with CSLA and blazor including their grid.

**Reply by fundez** [reply]
We also own devexpress. the only issue is currently devexpress grids doesn't support batch editing. You need to click an edit button on each row to start editing. Though they are planning to enable this feature in December.

**Reply by fundez** [reply]
We also own devexpress. the only issue is currently devexpress grids doesn't support batch editing. You need to click an edit button on each row to start editing. Though they are planning to enable this feature in December.

**Reply by rockfordlhotka** [reply]
In terms of workarounds, I don't know, but can provide info around the way CSLA works.

CSLA creates domain instances like this:

1. var newObject = Activator.Create(businessType)
2. If newObject implements IUseApplicationContext, set ApplicationContext
3. Done

I'm guessing that the datagrid control is doing step 1 (???)

Obviously it can't know to do step 2, and so you need to handle some "object created" or "row created" event in the datagrid to do step 2.

_Ideally_ the datagrid would allow _you_ to create the instance of the object for each row, in which case you could use the CSLA helper method to create the BO instance, but failing that you'll need to interact with `IUseApplicationContext` as a separate step.

**Reply by fundez** [reply]
Thanks Rocky, I ended up going back to CSLA 5 while I wait for Telerik or devpress to release updates.  Luckily v5 supports the Idataportal so I don't need to remove dataportal calls so it will make upgrading to 7 later much easier.
